tests/LSP6KeyManager/tests/LSP6ControlledToken.test.ts

```diff
@@ -0,0 +1,840 @@
+          ERC725YKeys.LSP6["AddressPermissions:Permissions"] +
+          context.accounts[0].address.substring(2);
+        const value = PERMISSIONS.CALL;
+        const payload = context.token.interface.encodeFunctionData(
+          "setData(bytes32,bytes)",
+          [key, value]
+        );
+
+        await context.keyManager.connect(context.accounts[2]).execute(payload);
+
+        expect(await context.token["getData(bytes32)"](key)).to.equal(value);
+      });
+
+      it("should add back ALL_PERMISSIONS of the address", async () => {
+        const key =
+          ERC725YKeys.LSP6["AddressPermissions:Permissions"] +
+          context.accounts[0].address.substring(2);
+        const value = ALL_PERMISSIONS;
+        const payload = context.token.interface.encodeFunctionData(
+          "setData(bytes32,bytes)",
+          [key, value]
+        );
+
+        await context.keyManager.connect(context.accounts[2]).execute(payload);
+
+        expect(await context.token["getData(bytes32)"](key)).to.equal(value);
+      });
+    });
+
+    describe("testing ADDPERMISSIONS permission", () => {
+      before(async () => {
+        await addControllerWithPermission(
+          context,
+          context.accounts[3],
+          ARRAY_LENGTH.FOUR,
+          ARRAY_INDEX.THREE,
+          PERMISSIONS.ADDPERMISSIONS
+        );
+      });
+
+      it("should add the new controller without changing other controllers", async () => {
+        // Check that a new controller was added and other controllers remained intact
+        const keys = [
+          ERC725YKeys.LSP6["AddressPermissions[]"].length,
+          ERC725YKeys.LSP6["AddressPermissions[]"].index + ARRAY_INDEX.ZERO,
+          ERC725YKeys.LSP6["AddressPermissions[]"].index + ARRAY_INDEX.ONE,
+          ERC725YKeys.LSP6["AddressPermissions[]"].index + ARRAY_INDEX.TWO,
+          ERC725YKeys.LSP6["AddressPermissions[]"].index + ARRAY_INDEX.THREE,
+          ERC725YKeys.LSP6["AddressPermissions:Permissions"] +
+            context.accounts[0].address.substring(2),
+          ERC725YKeys.LSP6["AddressPermissions:Permissions"] +
+            context.accounts[1].address.substring(2),
+          ERC725YKeys.LSP6["AddressPermissions:Permissions"] +
+            context.accounts[2].address.substring(2),
+          ERC725YKeys.LSP6["AddressPermissions:Permissions"] +
+            context.accounts[3].address.substring(2),
+        ];
+        const values = [
+          ARRAY_LENGTH.FOUR,
+          context.accounts[0].address,
+          context.accounts[1].address,
+          context.accounts[2].address,
+          context.accounts[3].address,
+          ALL_PERMISSIONS,
+          PERMISSIONS.CHANGEOWNER,
+          PERMISSIONS.CHANGEPERMISSIONS,
+          PERMISSIONS.ADDPERMISSIONS,
+        ];
+
+        expect(
+          await context.token["getData(bytes32[])"](keys)
+        ).to.be.deep.equal(values);
+      });
+
+      it("should be allowed to add permissions", async () => {
```

Comment:
I suggest this wording, to make it more clear.

```suggestion
      it("should be allowed to add a new controller address with some permissions", async () => {
```